kythe/go/serving/tools/kythefs/kythefs.go

```diff
@@ -0,0 +1,401 @@
+	//     "foo"               | "bar"
+	//     "foo/bar"           | "baz"
+	//
+	NextDirComponent string
+}
+
+// True if 'rs' contains a 'NextDirComponent' resolution.
+func hasDirComponent(rs []FilepathResolution) bool {
+	for _, r := range rs {
+		if r.NextDirComponent != "" {
+			return true
+		}
+	}
+	return false
+}
+
+// Returns alternative resolutions if the path is ambiguous.
+// Ambiguity comes from two sources:
+//
+//     a) The queried path points into a prefix of some corpus+root vfs path.
+//
+//     b) There are overlapping corpus+root+path vfs paths. If happens, you
+//        likely need to adjust the extractor's vname mapping config.
+//
+func (me *KytheFs) ResolveFilepath(path string) ([]FilepathResolution, error) {
+	req := ftpb.CorpusRootsRequest{}
+	cr, err := me.FileTree.CorpusRoots(me.Context, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set to dedup next-dirs, as they can be common against multiple corpus+root
+	// pairs.
+	var nextDirs map[string]bool = make(map[string]bool)
+
+	// Among multiple possible corpus+root pairs, prefer the longest ones.
+	// We could alternatively check existence of the path with API calls, but
+	// this situation shouldn't normally arise unless extraction config is broken.
+	longestCorpusRootForTicket := ""
+	var ticketResolution *FilepathResolution
+
+	sep := string(os.PathSeparator)
+
+	for _, corpus := range cr.Corpus {
+		for _, root := range corpus.Root {
+			crPath := filepath.Join(corpus.Name, root)
+			if path == "" {
+				// List root dirs, which are the first components of corpus names.
+				parts := filepath.SplitList(corpus.Name)
+				if len(parts) > 0 {
+					nextDirs[parts[0]] = true
+				}
+			} else {
+				crRemain := strings.TrimPrefix(crPath, path)
+				vfsRemain := strings.TrimPrefix(path, crPath)
+				if len(vfsRemain) < len(path) &&
+					(vfsRemain == "" || vfsRemain[:1] == sep) &&
+					len(crPath) > len(longestCorpusRootForTicket) {
+
+					longestCorpusRootForTicket = crPath
+					// Points inside corpus+root, match using the ticket.
+					var p string = "" // Exact corpus+root match.
+					if vfsRemain != "" {
+						p = vfsRemain[1:] // Additional kythe path.
+					}
+					ticketResolution = &FilepathResolution{KytheURI: &kytheuri.URI{
+						Corpus: corpus.Name,
+						Root:   root,
+						Path:   p,
+					}}
+				} else if len(crRemain) < len(crPath) && crRemain[:1] == sep {
+					// Queried path is a proper prefix.
+					parts := strings.Split(crRemain[1:], sep)
+					if len(parts) > 0 {
+						nextDirs[parts[0]] = true
+					}
+				}
+			}
+		}
+	}
+
+	var results []FilepathResolution
+	for k, _ := range nextDirs {
+		res := FilepathResolution{NextDirComponent: k}
+		results = append(results, res)
+	}
+	if ticketResolution != nil {
+		results = append(results, *ticketResolution)
+	}
+	return results, nil
+}
+
+// True if the given Kythe URI corresponds to a directory.
+func (me *KytheFs) IsDirectory(uri kytheuri.URI) (bool, error) {
```

Comment:
This really checks if the URI is **not** a known file rather than **is** a known directory.  You could also use the `FileTree` API to request the parent directory and actually check if the path is a directory.